timemachine/cpp/src/nonbonded_mol_energy.cu

```diff
@@ -0,0 +1,173 @@
+    }
+
+    int offset = 0;
+    // setup the mol idxs and the atom idxs
+    std::vector<int> mol_offsets(num_mols + 1);
+    std::vector<int> mol_idxs(num_grouped_atoms);
+    std::vector<int> atom_idxs(num_grouped_atoms);
+    for (int i = 0; i < num_mols; i++) {
+        std::vector<int> atoms = group_idxs[i];
+        // ASSUMES MOLS ARE MADE UP CONTIGIOUS INDICES IE mols[0] = [0, 1, 2], mols[1] = [3, 2], etc
+        // IF mols[0] = [5, 7], mols[1] = [0, 8], THIS WON'T WORK.
+        // If this is not the case need a complete mapping of atom_idx to mol_idx
+        // Sort the atom indices from smallest to largest so that you can know the range of indices in the molecule
+        std::sort(atoms.begin(), atoms.end());
+        int num_atoms = atoms.size();
+        mol_offsets[i] = offset;
+        for (int j = 0; j < num_atoms; j++) {
+            mol_idxs[offset + j] = i;
+            atom_idxs[offset + j] = atoms[j];
+        }
+        offset += num_atoms;
+    }
+    mol_offsets[num_mols] = offset;
+
+    return std::array<std::vector<int>, 3>({atom_idxs, mol_idxs, mol_offsets});
+}
+
+template <typename RealType>
+NonbondedMolEnergyPotential<RealType>::NonbondedMolEnergyPotential(
+    const int N, const std::vector<std::vector<int>> target_mols, const double beta, const double cutoff)
+    : N_(N), num_target_mols_(target_mols.size()), beta_(static_cast<RealType>(beta)),
+      cutoff_squared_(static_cast<RealType>(cutoff * cutoff)) {
+    verify_group_idx(N_, target_mols);
+
+    std::array<std::vector<int>, 3> target_flattened_groups = prepare_group_idxs_for_gpu(target_mols);
+
+    d_target_atom_idxs_.realloc(target_flattened_groups[0].size());
+    d_target_atom_idxs_.copy_from(&target_flattened_groups[0][0]);
+
+    d_target_mol_idxs_.realloc(target_flattened_groups[1].size());
+    d_target_mol_idxs_.copy_from(&target_flattened_groups[1][0]);
+
+    d_target_mol_offsets_.realloc(target_flattened_groups[2].size());
+    d_target_mol_offsets_.copy_from(&target_flattened_groups[2][0]);
+
+    // Need per atom energies per block to account for accumulating between different x blocks
+    // that have the same y block index
+    d_atom_energy_buffer_.realloc(d_target_atom_idxs_.length * BLOCK_SIZE);
+};
+
+template <typename RealType>
+void NonbondedMolEnergyPotential<RealType>::mol_energies_device(
+    const int N,
+    const int target_mols,
+    const double *d_coords,      // [N, 3]
+    const double *d_params,      // [N, PARAMS_PER_ATOM]
+    const double *d_box,         // [3, 3]
+    __int128 *d_output_energies, // [target_mols]
+    cudaStream_t stream) {
+    if (N != N_) {
+        throw std::runtime_error("N != N_");
+    }
+
+    if (target_mols != num_target_mols_) {
+        throw std::runtime_error("target_mols != num_target_mols_");
+    }
+
+    const int tpb = THREADS_PER_BLOCK;
+    // Brute force num_target_atoms X all atoms
+    dim3 dimGrid(BLOCK_SIZE, static_cast<int>(d_target_atom_idxs_.length), 1);
```

Comment:
Having the Y part of the dim being target atoms blows up once we get above 60k atoms. Can work around this, but not sure if it is worth doing in this PR.

Reply:
Inclined to leave this to when we do a neighborlist implementation of per mol energies, which would resolve this issue. Between now and then, don't forsee cases of a system with more than 60k atoms that need to be evaluated on a per mol basis.

Reply:
Is the limitation due to max size of `blockDim.y`? Can we just do multiple round-trips for now? We can definitely have systems over 60k atoms.

Reply:
Correct. Yes, we can add a loop inside and use a fixed block size. Will add the tests then implement this.

Reply:
Done in https://github.com/proteneer/timemachine/pull/1198/commits/b03a0f27ae381d34b3b488a1d3ee8c9acf89a923

Seems about 10% faster as well, which is a nice benefit.